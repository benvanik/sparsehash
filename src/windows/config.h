/* Handmade config.h for windows.  This is intended to be used with VC++,
   which does not use the unix-based './configure' configuration tool.
   These settings were taken from this patch, by Tobias Polzin:
      http://code.google.com/p/google-sparsehash/issues/attachment?aid=-71112522921222597&name=patch-sparsehash-0.6-win

   Note: when updating this file, you should include values for all the
         variables in src/config.h.include.
*/

/***
 *** Feel free to change these values to whatever suits your fancy
 ***/
#define GOOGLE_NAMESPACE  google

/***
 *** These are windows-specific; not from config.h.include
 ***/
#define HAVE_WINDOWS_H  1
#define snprintf  _snprintf
#define _DEFINE_DEPRECATED_HASH_CLASSES  0

/***
 *** Do not change these values unless you know they are wrong for your system
 ***/
#define HASH_FUN_H <google/sparsehash/stl_hash_fun.h>
#define HASH_MAP_H <hash_map>
#define HASH_NAMESPACE  stdext
/* #undef HAVE_UINT16_T */
/* #undef HAVE_U_INT16_T */
#define HAVE___UINT16  1
#define HAVE_LONG_LONG  1
/* #undef HAVE_SYS_TYPES_H */
/* #undef HAVE_STDINT_H */
/* #undef HAVE_INTTYPES_H */
#define HAVE_MEMCPY  1
#define STL_NAMESPACE  std
#define _END_GOOGLE_NAMESPACE_    };
#define _START_GOOGLE_NAMESPACE_  namespace GOOGLE_NAMESPACE {

/***
 *** These are only needed for the unittests and could go outside config.h
 ***/
/* #undef HAVE_SYS_RESOURCE_H */
/* #undef HAVE_UNISTD_H */
/* #undef HAVE_SYS_TIME_H */
/* #undef HAVE_SYS_UTSNAME_H */
#define HAVE_HASH_MAP  1
/* #undef HAVE_GOOGLE_MALLOC_EXTENSION_H */
