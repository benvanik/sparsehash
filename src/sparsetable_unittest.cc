// Copyright (c) 2005, Google Inc.
// All rights reserved.
// 
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
// 
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
// 
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

// ---
// Author: Craig Silverstein
//
// This tests <google/sparsetable>
//
// Since sparsetable is templatized, it's important that we test every
// function in every class in this file -- not just to see if it
// works, but even if it compiles.

#include <google/sparsehash/config.h>
#include <string>
#include <stdio.h>
#include <string.h>         // for memcmp()
#include <stdlib.h>         // defines unlink() on some windows platforms(?)
#ifdef HAVE_UNISTD_H
#include <unistd.h>         // for unlink()
#include <string>
#endif
#include <google/sparsetable>

using STL_NAMESPACE::string;
using GOOGLE_NAMESPACE::sparsetable;

// Many sparsetable operations return a size_t.  Rather than have to
// use PRIuS everywhere, we'll just cast to a "big enough" value.
#define UL(x)    ( static_cast<unsigned long>(x) )


static char outbuf[10240];       // big enough for these tests
static char* out = outbuf;       // where to write next
#define LEFT (outbuf + sizeof(outbuf) - out)

#define TEST(cond)  out += snprintf(out, LEFT, #cond "? %s\n", \
                                    (cond) ? "yes" : "no");

// replaces buf, in-place, with a version of buf that lacks \r's.
static int strip_cr(char* buf, int len) {
  char *r, *w;
  for ( r = buf, w = buf; r - buf < len; r++ ) {
    if (*r != '\r')
      *w++ = *r;
  }
  return w - buf;
}

inline string as_string(int n) {
  const int N = 64;
  char buf[N];
  snprintf(buf, N, "%d", n);
  return string(buf);
}

// Test sparsetable with a POD type, int.
void test_int() {
  out += snprintf(out, LEFT, "int test\n");
  sparsetable<int> x(7), y(70), z;
  x.set(4, 10);
  y.set(12, -12);
  y.set(47, -47);
  y.set(48, -48);
  y.set(49, -49);

  const sparsetable<int> constx(x);
  const sparsetable<int> consty(y);

  // ----------------------------------------------------------------------
  // Test the plain iterators

  for ( sparsetable<int>::iterator it = x.begin(); it != x.end(); ++it ) {
    out += snprintf(out, LEFT, "x[%lu]: %d\n", UL(it - x.begin()), int(*it));
  }
  for ( sparsetable<int>::const_iterator it = x.begin(); it != x.end(); ++it ) {
    out += snprintf(out, LEFT, "x[%lu]: %d\n", UL(it - x.begin()), *it);
  }
  for ( sparsetable<int>::reverse_iterator it = x.rbegin(); it != x.rend(); ++it ) {
    out += snprintf(out, LEFT, "x[%lu]: %d\n", UL(x.rend()-1 - it), int(*it));
  }
  for ( sparsetable<int>::const_reverse_iterator it = constx.rbegin(); it != constx.rend(); ++it ) {
    out += snprintf(out, LEFT, "x[%lu]: %d\n", UL(constx.rend()-1 - it), *it);
  }
  for ( sparsetable<int>::iterator it = z.begin(); it != z.end(); ++it ) {
    out += snprintf(out, LEFT, "z[%lu]: %d\n", UL(it - z.begin()), int(*it));
  }

  {                                          // array version
    out += snprintf(out, LEFT, "x[3]: %d\n", int(x[3]));
    out += snprintf(out, LEFT, "x[4]: %d\n", int(x[4]));
    out += snprintf(out, LEFT, "x[5]: %d\n", int(x[5]));
  }
  {
    sparsetable<int>::iterator it;           // non-const version
    out += snprintf(out, LEFT, "x[4]: %d\n", int(x.begin()[4]));
    it = x.begin() + 4;          // should point to the non-zero value
    out += snprintf(out, LEFT, "x[4]: %d\n", int(*it));
    it--;
    --it;
    it += 5;
    it -= 2;
    it++;
    ++it;
    it = it - 3;
    it = 1 + it;                 // now at 5
    out += snprintf(out, LEFT, "x[3]: %d\n", int(it[-2]));
    out += snprintf(out, LEFT, "x[4]: %d\n", int(it[-1]));
    *it = 55;
    out += snprintf(out, LEFT, "x[5]: %d\n", int(it[0]));
    out += snprintf(out, LEFT, "x[5]: %d\n", int(*it));
    int *x6 = &(it[1]);
    *x6 = 66;
    out += snprintf(out, LEFT, "x[6]: %d\n", int(*(it + 1)));
    // Let's test comparitors as well
    TEST(it == it);
    TEST(!(it != it));
    TEST(!(it < it));
    TEST(!(it > it));
    TEST(it <= it);
    TEST(it >= it);

    sparsetable<int>::iterator it_minus_1 = it - 1;
    TEST(!(it == it_minus_1));
    TEST(it != it_minus_1);
    TEST(!(it < it_minus_1));
    TEST(it > it_minus_1);
    TEST(!(it <= it_minus_1));
    TEST(it >= it_minus_1);
    TEST(!(it_minus_1 == it));
    TEST(it_minus_1 != it);
    TEST(it_minus_1 < it);
    TEST(!(it_minus_1 > it));
    TEST(it_minus_1 <= it);
    TEST(!(it_minus_1 >= it));

    sparsetable<int>::iterator it_plus_1 = it + 1;
    TEST(!(it == it_plus_1));
    TEST(it != it_plus_1);
    TEST(it < it_plus_1);
    TEST(!(it > it_plus_1));
    TEST(it <= it_plus_1);
    TEST(!(it >= it_plus_1));
    TEST(!(it_plus_1 == it));
    TEST(it_plus_1 != it);
    TEST(!(it_plus_1 < it));
    TEST(it_plus_1 > it);
    TEST(!(it_plus_1 <= it));
    TEST(it_plus_1 >= it);
  }
  {
    sparsetable<int>::const_iterator it;    // const version
    out += snprintf(out, LEFT, "x[4]: %d\n", int(x.begin()[4]));
    it = x.begin() + 4;          // should point to the non-zero value
    out += snprintf(out, LEFT, "x[4]: %d\n", *it);
    it--;
    --it;
    it += 5;
    it -= 2;
    it++;
    ++it;
    it = it - 3;
    it = 1 + it;                 // now at 5
    out += snprintf(out, LEFT, "x[3]: %d\n", it[-2]);
    out += snprintf(out, LEFT, "x[4]: %d\n", it[-1]);
    out += snprintf(out, LEFT, "x[5]: %d\n", *it);
    out += snprintf(out, LEFT, "x[6]: %d\n", *(it + 1));
    // Let's test comparitors as well
    TEST(it == it);
    TEST(!(it != it));
    TEST(!(it < it));
    TEST(!(it > it));
    TEST(it <= it);
    TEST(it >= it);

    sparsetable<int>::const_iterator it_minus_1 = it - 1;
    TEST(!(it == it_minus_1));
    TEST(it != it_minus_1);
    TEST(!(it < it_minus_1));
    TEST(it > it_minus_1);
    TEST(!(it <= it_minus_1));
    TEST(it >= it_minus_1);
    TEST(!(it_minus_1 == it));
    TEST(it_minus_1 != it);
    TEST(it_minus_1 < it);
    TEST(!(it_minus_1 > it));
    TEST(it_minus_1 <= it);
    TEST(!(it_minus_1 >= it));

    sparsetable<int>::const_iterator it_plus_1 = it + 1;
    TEST(!(it == it_plus_1));
    TEST(it != it_plus_1);
    TEST(it < it_plus_1);
    TEST(!(it > it_plus_1));
    TEST(it <= it_plus_1);
    TEST(!(it >= it_plus_1));
    TEST(!(it_plus_1 == it));
    TEST(it_plus_1 != it);
    TEST(!(it_plus_1 < it));
    TEST(it_plus_1 > it);
    TEST(!(it_plus_1 <= it));
    TEST(it_plus_1 >= it);
  }

  TEST(x.begin() == x.begin() + 1 - 1);
  TEST(x.begin() < x.end());
  TEST(z.begin() < z.end());
  TEST(z.begin() <= z.end());
  TEST(z.begin() == z.end());


  // ----------------------------------------------------------------------
  // Test the non-empty iterators

  for ( sparsetable<int>::nonempty_iterator it = x.nonempty_begin(); it != x.nonempty_end(); ++it ) {
    out += snprintf(out, LEFT, "x[??]: %d\n", *it);
  }
  for ( sparsetable<int>::const_nonempty_iterator it = y.nonempty_begin(); it != y.nonempty_end(); ++it ) {
    out += snprintf(out, LEFT, "y[??]: %d\n", *it);
  }
  for ( sparsetable<int>::reverse_nonempty_iterator it = y.nonempty_rbegin(); it != y.nonempty_rend(); ++it ) {
    out += snprintf(out, LEFT, "y[??]: %d\n", *it);
  }
  for ( sparsetable<int>::const_reverse_nonempty_iterator it = consty.nonempty_rbegin(); it != consty.nonempty_rend(); ++it ) {
    out += snprintf(out, LEFT, "y[??]: %d\n", *it);
  }
  for ( sparsetable<int>::nonempty_iterator it = z.nonempty_begin(); it != z.nonempty_end(); ++it ) {
    out += snprintf(out, LEFT, "z[??]: %d\n", *it);
  }

  {
    sparsetable<int>::nonempty_iterator it;           // non-const version
    out += snprintf(out, LEFT, "first non-empty y: %d\n", *y.nonempty_begin());
    out += snprintf(out, LEFT, "first non-empty x: %d\n", *x.nonempty_begin());
    it = x.nonempty_begin();
    ++it;                        // should be at end
    --it;
    out += snprintf(out, LEFT, "first non-empty x: %d\n", *it++);
    it--;
    out += snprintf(out, LEFT, "first non-empty x: %d\n", *it++);
  }
  {
    sparsetable<int>::const_nonempty_iterator it;           // non-const version
    out += snprintf(out, LEFT, "first non-empty y: %d\n", *y.nonempty_begin());
    out += snprintf(out, LEFT, "first non-empty x: %d\n", *x.nonempty_begin());
    it = x.nonempty_begin();
    ++it;                        // should be at end
    --it;
    out += snprintf(out, LEFT, "first non-empty x: %d\n", *it++);
    it--;
    out += snprintf(out, LEFT, "first non-empty x: %d\n", *it++);
  }

  TEST(x.begin() == x.begin() + 1 - 1);
  TEST(z.begin() != z.end());

  // ----------------------------------------------------------------------
  // Test sparsetable functions
  out += snprintf(out, LEFT, "x has %lu/%lu buckets, "
                  "y %lu/%lu, z %lu/%lu\n",
                  UL(x.num_nonempty()), UL(x.size()),
                  UL(y.num_nonempty()), UL(y.size()),
                  UL(z.num_nonempty()), UL(z.size()));
  
  y.resize(48);              // should get rid of 48 and 49
  y.resize(70);              // 48 and 49 should still be gone
  out += snprintf(out, LEFT, "y shrank and grew: it's now %lu/%lu\n",
                  UL(y.num_nonempty()), UL(y.size()));
  out += snprintf(out, LEFT, "y[12] = %d, y.get(12) = %d\n", int(y[12]), y.get(12));
  y.erase(12);
  out += snprintf(out, LEFT, "y[12] cleared.  y now %lu/%lu.  "
                  "y[12] = %d, y.get(12) = %d\n",
                  UL(y.num_nonempty()), UL(y.size()), int(y[12]), y.get(12));

  swap(x, y);

  y.clear();
  TEST(y == z);

  y.resize(70);
  for ( int i = 10; i < 40; ++i )
    y[i] = -i;
  y.erase(y.begin() + 15, y.begin() + 30);
  y.erase(y.begin() + 34);
  y.erase(12);
  y.resize(38);
  y.resize(10000);
  y[9898] = -9898;
  for ( sparsetable<int>::const_iterator it = y.begin(); it != y.end(); ++it ) {
    if ( y.test(it) )
      out += snprintf(out, LEFT, "y[%lu] is set\n", UL(it - y.begin()));
  }
  out += snprintf(out, LEFT, "That's %lu set buckets\n", UL(y.num_nonempty()));

  out += snprintf(out, LEFT, "Starting from y[32]...\n");
  for ( sparsetable<int>::const_nonempty_iterator it = y.get_iter(32);
        it != y.nonempty_end(); ++it )
    out += snprintf(out, LEFT, "y[??] = %d\n", *it);

  out += snprintf(out, LEFT, "From y[32] down...\n");
  for ( sparsetable<int>::nonempty_iterator it = y.get_iter(32);
        it != y.nonempty_begin(); )
    out += snprintf(out, LEFT, "y[??] = %d\n", *--it);

  // ----------------------------------------------------------------------
  // Test I/O
  string filestr = "/tmp/#sparsetable.test";
  const char *file = filestr.c_str();
  FILE *fp = fopen(file, "wb");
  if ( fp == NULL ) {
    // maybe we can't write to /tmp/.  Try the current directory
    file = "#sparsetable.test";
    fp = fopen(file, "wb");
  }
  if ( fp == NULL ) {
    out += snprintf(out, LEFT, "Can't open %s, skipping disk write...\n", file);
  } else {
    y.write_metadata(fp);          // only write meta-information
    y.write_nopointer_data(fp);
    fclose(fp);
  }
  fp = fopen(file, "rb");
  if ( fp == NULL ) {
    out += snprintf(out, LEFT, "Can't open %s, skipping disk read...\n", file);
  } else {
    sparsetable<int> y2;
    y2.read_metadata(fp);
    y2.read_nopointer_data(fp);
    fclose(fp);

    for ( sparsetable<int>::const_iterator it = y2.begin(); it != y2.end(); ++it ) {
      if ( y2.test(it) )
        out += snprintf(out, LEFT, "y2[%lu] is %d\n", UL(it - y2.begin()), *it);
    }
    out += snprintf(out, LEFT, "That's %lu set buckets\n", UL(y2.num_nonempty()));
  }
  unlink(file);
}

// Test sparsetable with a non-POD type, std::string
void test_string() {
  out += snprintf(out, LEFT, "string test\n");
  sparsetable<string> x(7), y(70), z;
  x.set(4, "foo");
  y.set(12, "orange");
  y.set(47, "grape");
  y.set(48, "pear");
  y.set(49, "apple");

  // ----------------------------------------------------------------------
  // Test the plain iterators

  for ( sparsetable<string>::iterator it = x.begin(); it != x.end(); ++it ) {
    out += snprintf(out, LEFT, "x[%lu]: %s\n",
                    UL(it - x.begin()), static_cast<string>(*it).c_str());
  }
  for ( sparsetable<string>::iterator it = z.begin(); it != z.end(); ++it ) {
    out += snprintf(out, LEFT, "z[%lu]: %s\n",
                    UL(it - z.begin()), static_cast<string>(*it).c_str());
  }

  TEST(x.begin() == x.begin() + 1 - 1);
  TEST(x.begin() < x.end());
  TEST(z.begin() < z.end());
  TEST(z.begin() <= z.end());
  TEST(z.begin() == z.end());

  // ----------------------------------------------------------------------
  // Test the non-empty iterators
    for ( sparsetable<string>::nonempty_iterator it = x.nonempty_begin(); it != x.nonempty_end(); ++it ) {
    out += snprintf(out, LEFT, "x[??]: %s\n", it->c_str());
  }
  for ( sparsetable<string>::const_nonempty_iterator it = y.nonempty_begin(); it != y.nonempty_end(); ++it ) {
    out += snprintf(out, LEFT, "y[??]: %s\n", it->c_str());
  }
  for ( sparsetable<string>::nonempty_iterator it = z.nonempty_begin(); it != z.nonempty_end(); ++it ) {
    out += snprintf(out, LEFT, "z[??]: %s\n", it->c_str());
  }

  // ----------------------------------------------------------------------
  // Test sparsetable functions
  out += snprintf(out, LEFT, "x has %lu/%lu buckets, y %lu/%lu, z %lu/%lu\n",
                  UL(x.num_nonempty()), UL(x.size()),
                  UL(y.num_nonempty()), UL(y.size()),
                  UL(z.num_nonempty()), UL(z.size()));

  y.resize(48);              // should get rid of 48 and 49
  y.resize(70);              // 48 and 49 should still be gone
  out += snprintf(out, LEFT, "y shrank and grew: it's now %lu/%lu\n",
                  UL(y.num_nonempty()), UL(y.size()));
  out += snprintf(out, LEFT, "y[12] = %s, y.get(12) = %s\n",
                  static_cast<string>(y[12]).c_str(), y.get(12).c_str());
  y.erase(12);
  out += snprintf(out, LEFT, "y[12] cleared.  y now %lu/%lu.  "
                  "y[12] = %s, y.get(12) = %s\n",
                  UL(y.num_nonempty()), UL(y.size()),
                  static_cast<string>(y[12]).c_str(),
                  static_cast<string>(y.get(12)).c_str());
  swap(x, y);

  y.clear();
  TEST(y == z);

  y.resize(70);
  for ( int i = 10; i < 40; ++i )
    y.set(i, as_string(-i));
  y.erase(y.begin() + 15, y.begin() + 30);
  y.erase(y.begin() + 34);
  y.erase(12);
  y.resize(38);
  y.resize(10000);
  y.set(9898, as_string(-9898));
  for ( sparsetable<string>::const_iterator it = y.begin(); it != y.end(); ++it ) {
    if ( y.test(it) )
      out += snprintf(out, LEFT, "y[%lu] is set\n", UL(it - y.begin()));
  }
  out += snprintf(out, LEFT, "That's %lu set buckets\n", UL(y.num_nonempty()));

  out += snprintf(out, LEFT, "Starting from y[32]...\n");
  for ( sparsetable<string>::const_nonempty_iterator it = y.get_iter(32);
        it != y.nonempty_end(); ++it )
    out += snprintf(out, LEFT, "y[??] = %s\n", it->c_str());

  out += snprintf(out, LEFT, "From y[32] down...\n");
  for ( sparsetable<string>::nonempty_iterator it = y.get_iter(32);
        it != y.nonempty_begin(); )
    out += snprintf(out, LEFT, "y[??] = %s\n", (*--it).c_str());
}

int main(int argc, char **argv) {          // though we ignore the args
  test_int();
  test_string();
  // Finally, check to see if our output (in out) is what it's supposed to be.
  char expected[sizeof(outbuf) + 10];
  string filename = (string(getenv("srcdir") ? getenv("srcdir") : ".") +
                     "/src/sparsetable_unittest.expected");
  FILE* expected_fp = fopen(filename.c_str(), "rb");
  if ( expected_fp == NULL ) {
    fprintf(stderr, "Can't judge test: can't load 'golden' file '%s'.\n",
            filename.c_str());
    return 2;
  }
  int r = fread(expected, 1, sizeof(expected) - 1, expected_fp);
  r = strip_cr(expected, r);
  if ( r != out - outbuf ||             // output not the same size
       memcmp(outbuf, expected, r) ) {  // or bytes differed
    fprintf(stderr, "TESTS FAILED\n\nEXPECTED:\n\n%s\n\nACTUAL:\n\n%s\n\n",
            expected, outbuf);
    return 1;
  } else {
    printf("All tests passed.\n");
    return 0;
  }
}
