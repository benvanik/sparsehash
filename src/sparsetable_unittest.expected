int test
x[0]: 0
x[1]: 0
x[2]: 0
x[3]: 0
x[4]: 10
x[5]: 0
x[6]: 0
x[0]: 0
x[1]: 0
x[2]: 0
x[3]: 0
x[4]: 10
x[5]: 0
x[6]: 0
x[6]: 0
x[5]: 0
x[4]: 10
x[3]: 0
x[2]: 0
x[1]: 0
x[0]: 0
x[6]: 0
x[5]: 0
x[4]: 10
x[3]: 0
x[2]: 0
x[1]: 0
x[0]: 0
x[3]: 0
x[4]: 10
x[5]: 0
x[4]: 10
x[4]: 10
x[3]: 0
x[4]: 10
x[5]: 55
x[5]: 55
x[6]: 66
it == it? yes
!(it != it)? yes
!(it < it)? yes
!(it > it)? yes
it <= it? yes
it >= it? yes
!(it == it_minus_1)? yes
it != it_minus_1? yes
!(it < it_minus_1)? yes
it > it_minus_1? yes
!(it <= it_minus_1)? yes
it >= it_minus_1? yes
!(it_minus_1 == it)? yes
it_minus_1 != it? yes
it_minus_1 < it? yes
!(it_minus_1 > it)? yes
it_minus_1 <= it? yes
!(it_minus_1 >= it)? yes
!(it == it_plus_1)? yes
it != it_plus_1? yes
it < it_plus_1? yes
!(it > it_plus_1)? yes
it <= it_plus_1? yes
!(it >= it_plus_1)? yes
!(it_plus_1 == it)? yes
it_plus_1 != it? yes
!(it_plus_1 < it)? yes
it_plus_1 > it? yes
!(it_plus_1 <= it)? yes
it_plus_1 >= it? yes
x[4]: 10
x[4]: 10
x[3]: 0
x[4]: 10
x[5]: 55
x[6]: 66
it == it? yes
!(it != it)? yes
!(it < it)? yes
!(it > it)? yes
it <= it? yes
it >= it? yes
!(it == it_minus_1)? yes
it != it_minus_1? yes
!(it < it_minus_1)? yes
it > it_minus_1? yes
!(it <= it_minus_1)? yes
it >= it_minus_1? yes
!(it_minus_1 == it)? yes
it_minus_1 != it? yes
it_minus_1 < it? yes
!(it_minus_1 > it)? yes
it_minus_1 <= it? yes
!(it_minus_1 >= it)? yes
!(it == it_plus_1)? yes
it != it_plus_1? yes
it < it_plus_1? yes
!(it > it_plus_1)? yes
it <= it_plus_1? yes
!(it >= it_plus_1)? yes
!(it_plus_1 == it)? yes
it_plus_1 != it? yes
!(it_plus_1 < it)? yes
it_plus_1 > it? yes
!(it_plus_1 <= it)? yes
it_plus_1 >= it? yes
x.begin() == x.begin() + 1 - 1? yes
x.begin() < x.end()? yes
z.begin() < z.end()? no
z.begin() <= z.end()? yes
z.begin() == z.end()? yes
x[??]: 10
x[??]: 55
x[??]: 66
y[??]: -12
y[??]: -47
y[??]: -48
y[??]: -49
y[??]: -49
y[??]: -48
y[??]: -47
y[??]: -12
y[??]: -49
y[??]: -48
y[??]: -47
y[??]: -12
first non-empty y: -12
first non-empty x: 10
first non-empty x: 10
first non-empty x: 10
first non-empty y: -12
first non-empty x: 10
first non-empty x: 10
first non-empty x: 10
x.begin() == x.begin() + 1 - 1? yes
z.begin() != z.end()? no
x has 3/7 buckets, y 4/70, z 0/0
y shrank and grew: it's now 2/70
y[12] = -12, y.get(12) = -12
y[12] cleared.  y now 1/70.  y[12] = 0, y.get(12) = 0
y == z? no
y[10] is set
y[11] is set
y[13] is set
y[14] is set
y[30] is set
y[31] is set
y[32] is set
y[33] is set
y[35] is set
y[36] is set
y[37] is set
y[9898] is set
That's 12 set buckets
Starting from y[32]...
y[??] = -32
y[??] = -33
y[??] = -35
y[??] = -36
y[??] = -37
y[??] = -9898
From y[32] down...
y[??] = -31
y[??] = -30
y[??] = -14
y[??] = -13
y[??] = -11
y[??] = -10
y2[10] is -10
y2[11] is -11
y2[13] is -13
y2[14] is -14
y2[30] is -30
y2[31] is -31
y2[32] is -32
y2[33] is -33
y2[35] is -35
y2[36] is -36
y2[37] is -37
y2[9898] is -9898
That's 12 set buckets
string test
x[0]: 
x[1]: 
x[2]: 
x[3]: 
x[4]: foo
x[5]: 
x[6]: 
x.begin() == x.begin() + 1 - 1? yes
x.begin() < x.end()? yes
z.begin() < z.end()? no
z.begin() <= z.end()? yes
z.begin() == z.end()? yes
x[??]: foo
y[??]: orange
y[??]: grape
y[??]: pear
y[??]: apple
x has 1/7 buckets, y 4/70, z 0/0
y shrank and grew: it's now 2/70
y[12] = orange, y.get(12) = orange
y[12] cleared.  y now 1/70.  y[12] = , y.get(12) = 
y == z? no
y[10] is set
y[11] is set
y[13] is set
y[14] is set
y[30] is set
y[31] is set
y[32] is set
y[33] is set
y[35] is set
y[36] is set
y[37] is set
y[9898] is set
That's 12 set buckets
Starting from y[32]...
y[??] = -32
y[??] = -33
y[??] = -35
y[??] = -36
y[??] = -37
y[??] = -9898
From y[32] down...
y[??] = -31
y[??] = -30
y[??] = -14
y[??] = -13
y[??] = -11
y[??] = -10
